### DISCUSSION TITLE
Publishing quarto content in another quarto project (e.g. quarto-project -> quarto-blog)

### DISCUSSION THREAD

**Discussion by sjdh** [quarto-dev/quarto-cli#11677] [Q&A] [answered]
### Description

How to export results from a quarto project into a stand-alone quarto blog?

I have a couple of projects with quarto content, as well as a quarto blog:

```
- project a
  - post-a1.qmd
  - post-a2.qmd
  - data, envs, results
- project 2
  - blog-b1.qmd
  - blog-b2.qmd
   - data, envs, results
- quarto-blog
   - post1.qmd
      - env 1
   - post2.qmd
       - env 2
```

The projects are stand-alone. They have their own environments and dependencies;
which may include large datasets and heavy compute. The quarto blog should not become dependent on these heavy projects.

Still I like to write about the projects in the blog, and include some code, results, figures and references.
Resources (code, figures, data) and project related quarto code are available in the project.
It will be efficient to write the blog posts within the project and then export the result to blog. 

Is there a workflow for that?

Quarto can render content to targets like pdf, html, and even mardown to be included in a hugo blog. 

I was thinking something like:

```
quarto project -> quarto-blog -> html
```

**Comment by mcanouil** [comment]
You'll have to come up yourself with a workflow to make that work based on what you want/need.
Some paths to explore:
- https://quarto.org/docs/projects/profiles.html
- https://quarto.org/docs/projects/code-execution.html#freeze
- https://quarto.org/docs/authoring/notebook-embed.html
- https://quarto.org/docs/authoring/includes.html
- iframe via `![my file](path/to/file.ext)`

**Reply by sjdh** [reply]
Profiles, freeze, notebook embed and include all execute code within the quarto blog. I would still need the projects as dependencies within the blog.



What I want is execute code outside the quarto blog (inside the project), and write the output (as markdown + assets?) to the source folder of the quarto blog.

**Reply by mcanouil** [reply]
Quarto's design for projects that won't change is that they are self-contained.
What you want is to go around that, so you need to build that workflow yourself: compiling notebooks, using freeze/cache, reusing/copying freeze/cache, etc.

Note that you could use profiles and different environment/dependencies to generate the freeze/cache.

**Reply by sjdh** [reply]
The workflow that I am trying now is to convert the qmd file in the project to a md file in the quarto-blog.

```
quarto render --output-dir --to md  quarto_blog/posts blog_post.qmd 
```

**Reply by sjdh** [reply]
That approach has the problem that quarto blog cannot use `.md` files as posts:

```bash
cd quarto_blog
quarto preview
ERROR: You must use the .qmd extension for documents with executable code.
```

**Reply by mcanouil** [reply]
> That approach has the problem that quarto blog cannot use .md files as posts:

That's not what the message says. Quarto supports "md" files.
You have code cells (executable code) which are code to be executed. Such code needs to be in a "qmd" file.

**Reply by sjdh** [reply]
Thanks for pointing that out!

After conversion to markdown with `quarto render --to md `, my code blocks look like this:

```md
 279   │ ``` {r}
 280   │ plot_trajectory(trajectories, 1)
 281   │ ```
``` 
Maybe quarto accepts the `.md` post if code blocks look like this?
```
 279   │ ```r
 280   │ plot_trajectory(trajectories, 1)
 281   │ ```
```

Should the `--to md` option not do this automatically?

**Reply by mcanouil** [reply]
The brackets around the language is Quarto's syntax to say "this is code to be executed", so using a code block without brackets is valid for "md" file in Quarto.

**Reply by sjdh** [reply]
> The brackets around the language is Quarto's syntax to say "this is code to be executed"

So `{r}` is quarto or `.qmd` syntax. 

Then should `quarto render --to md` not convert `{r}` to `r`? After all markdown does not have executable code.

**Reply by mcanouil** [reply]
You need to provide something actionable at some point.

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
title: "Quarto Playground"
format: md
---

```{r}
print("Hello, World!")
```

````

</td><td>

````md
# Quarto Playground


``` r
print("Hello, World!")
```

    [1] "Hello, World!"
````    

</td></tr>
</table>

**Reply by sjdh** [reply]
How about this? 

I'm using `sed` now to finish the conversion `quarto render --to md`  from qmd to markdown. With sed I change all executable code blocks `{r}` into just code blocks. 

What is a better way to do this? Can quarto not do this by itself?

**Reply by mcanouil** [reply]
As illustrated, Quarto with "md" as format won't have brackets so the issue is with your files/setup/all.
Again, I'm afraid if you don't provide something we can use to help you, my help will stop there unfortunately.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by sjdh** [reply]
Thanks. I did not get the illustration before your last comment. 

After seeing that quarto does the conversion from executable codeblocks to markdown codeblocks correct, tried with my example. Removing the `--output-dir` flag made it work. I have no minimal example to reproduce it. Maybe I sent it later. 

Thanks for your help. My question is solved. I will post the solution below.

**Comment by sjdh** [comment, marked as answer]
# Workflow "publishing from project to blog" 

### Problem
While working on a stand-alone project, I want to publish a post to my quarto blog from the project.
I do not want to make the project a dependency of the blog. 



### Solution
Keep all processing in the context of the project, including

- Execute code
- Generate figures

### Step 1 - Render post in the project

```
quarto render --output-dir <path to blog>/posts --to md  blog_post.qmd 
```

### Step 2 - Render the blog
```
quarto render 
```

### Possible bug
On my setup, Quarto executable blocks are not converted to markdown code blocks when I run 
```
quarto render --output-dir <path to blog>/posts--to md   blog_post.qmd 
```
Removing the `--output-dir` option solved this.
